src/server/pps/server/githook/githook.go

```diff
@@ -0,0 +1,159 @@
+package githook
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"path"
+	"strconv"
+
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/pps"
+	col "github.com/pachyderm/pachyderm/src/server/pkg/collection"
+	"github.com/pachyderm/pachyderm/src/server/pkg/ppsdb"
+	"github.com/pachyderm/pachyderm/src/server/pkg/util"
+
+	etcd "github.com/coreos/etcd/clientv3"
+	logrus "github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+	"gopkg.in/go-playground/webhooks.v3"
+	"gopkg.in/go-playground/webhooks.v3/github"
+)
+
+// GitHookPort specifies the port the server will listen on
+const GitHookPort = 999
+const apiVersion = "v1"
+
+// gitHookServer serves GetFile requests over HTTP
+type gitHookServer struct {
+	hook       *github.Webhook
+	client     *client.APIClient
+	etcdClient *etcd.Client
+	pipelines  col.Collection
+}
+
+// RunGitHookServer starts the webhook server
+func RunGitHookServer(address string, etcdAddress string, etcdPrefix string) error {
+	c, err := client.NewFromAddress(address)
+	if err != nil {
+		return err
+	}
+	etcdClient, err := etcd.New(etcd.Config{
+		Endpoints:   []string{etcdAddress},
+		DialOptions: client.EtcdDialOptions(),
+	})
+	if err != nil {
+		return err
+	}
+	hook := github.New(&github.Config{})
+	s := &gitHookServer{
+		hook,
+		c,
+		etcdClient,
+		ppsdb.Pipelines(etcdClient, etcdPrefix),
+	}
+
+	hook.RegisterEvents(
+		func(payload interface{}, header webhooks.Header) {
+			if err := s.HandlePush(payload, header); err != nil {
+				pl := payload.(github.PushPayload)
```

Comment:
Could we capture / check the `ok` parameter here? I'm guessing this library guarantees that the value it gives to this function will be of type `PushPayload` but given that it's not reflected in the type I think we should be defensive about it since if it's a different type this will segfault.

Reply:
You could also make `HandlePush` take a `github.PushPayload` so you only need to do the cast once.